File: src/javascript/algorithms/array/15.buySellStock/buySellStock.js
function buySellStock(stockPrices) {
    if(stockPrices.length < 2) return 0;

    let maxProfit = 0;
    let left = 0, right = 1;

    while(right < stockPrices.length) {
        if(stockPrices[right] > stockPrices[left]) {
            maxProfit = Math.max(maxProfit, stockPrices[right] - stockPrices[left]);
        } else {
            left = right; // // Found a lower price to buy a stock
        } 
        right++;
    }

    return maxProfit;
}

// Test case 1: Standard increasing/decreasing
console.log(buySellStock([8, 3, 6, 4, 7, 5])); // Expected: 4
// Test case 2: No profit possible (monotonically decreasing)
console.log(buySellStock([7, 6, 5, 4, 3, 2, 1])); // Expected: 0
// Test case 3: Increasing prices only
console.log(buySellStock([1, 2, 3, 4, 5, 6])); // Expected: 5
// Test case 4: Single price (not enough data)
console.log(buySellStock([7])); // Expected: 0
// Test case 5: Empty array
console.log(buySellStock([])); // Expected: 0
// Test case 6: Flat prices
console.log(buySellStock([5, 5, 5, 5, 5])); // Expected: 0

File: src/python/algorithms/array/15.buySellStock/buy_sell_stock.py
"""
Best Time to Buy and Sell Stock

You are given an array prices where prices[i] is the price of a given stock on the ith day.
Find the maximum profit you can achieve. You may complete at most one transaction.

Time Complexity: O(n)
Space Complexity: O(1)
"""

"""
One pass solution tracking minimum price.
TC: O(n), SC: O(1)
"""
def max_profit(prices):

    if not prices:
        return 0
    
    min_price = float('inf')
    max_profit = 0
    
    for price in prices:
        min_price = min(min_price, price)
        current_profit = price - min_price
        max_profit = max(max_profit, current_profit)
    
    return max_profit

"""
Two pointers solution.
TC: O(n), SC: O(1)
"""
def max_profit1(prices):

    if not prices:
        return 0

    left, right = 0, 1
    max_profit = 0

    while right in range(len(prices)):
        if prices[right] > prices[left]:
            max_profit = max(max_profit, prices[right]-prices[left])
        else:
            left = right // Jumpt to lower price
        right += 1
    return max_profit
    

# Test cases
if __name__ == "__main__":
    test_cases = [
        [7, 1, 5, 3, 6, 4],  # 5 (buy at 1, sell at 6)
        [7, 6, 4, 3, 1],     # 0 (no profit possible)
    ]
    
    for prices in test_cases:
        print(f"Prices: {prices}")
        print(f"Max profit: {max_profit(prices)}")
        print(f"Max profit: {max_profit1(prices)}")

        print()
